packages/ckeditor5-watchdog/src/editorwatchdog.ts

```diff
@@ -165,12 +164,24 @@ export default class EditorWatchdog<TEditor extends Editor = Editor> extends Wat
 				console.error( 'An error happened during the editor destroying.', err );
 			} )
 			.then( () => {
+				const existingRoots = Object.keys( this._data!.roots ).reduce( ( acc, rootName ) => {
+					acc[ rootName ] = '';
+
+					return acc;
+				}, {} as Record<string, string> );
+
+				const updatedConfig = {
+					...this._config,
+					plugins: this._config!.plugins || [],
+					_watchdogInitialData: this._data
+				};
+
+				updatedConfig.plugins!.push( EditorWatchdogInitPlugin as any );
+
 				if ( typeof this._elementOrData === 'string' ) {
-					return this.create( this._data, this._config, this._config!.context );
+					return this.create( existingRoots, updatedConfig, updatedConfig!.context );
 				} else {
-					const updatedConfig = Object.assign( {}, this._config, {
-						initialData: this._data
-					} );
+					updatedConfig.initialData = existingRoots;
```

Comment:
I am not sure if this will work correctly for multi-root editor, if you add some roots before there's editor restart. It may be that there will be letfovers in DOM after the crashed editor is destroyed. And then, after you restart, you will use only DOM elements for the initial roots (as they were saved in `this._elementOrData`). For the new roots, you will again create new DOM elements. Please check if that really happens.

In this case, you will need to update `this._elementOrData` with proper DOM elements as well :thinking:. Or fix it in a different way :D.

Reply:
Anyway, please remember, that `this._elementOrData` may still be an object with strings or with DOM elements, and it is supported for both multi-root and single-root editors.

Reply:
Yes, you are right and it doesn't work after adding/removing roots in the multi-root editor. As we discussed, it will be fixed in the next step (operations approach).